src/oauth/server.ts

```diff
@@ -0,0 +1,356 @@
+    logger.info('new client registered', {
+      clientId,
+      client_name,
+      redirect_uris,
+    });
+
+    res.json({
+      client_id: clientId,
+      client_secret: clientSecret,
+      client_name,
+      redirect_uris,
+    });
+  } catch (error: unknown) {
+    logger.error('failed to register client:', {
+      error: error instanceof Error ? error.message : 'Unknown error',
+      client: req.body.client_name,
+    });
+    res.status(400).json({ error: 'Invalid request' });
+  }
+};
+
+const authRouter = express.Router();
+authRouter.get('/.well-known/oauth-authorization-server', metadata);
+authRouter.post('/register', bodyParser.json(), registerClient);
+
+/*
+  Initiate the authorization code grant flow by redirecting to the upstream
+  authorization server.
+  
+  Step 1:
+  MCP client should invoke this endpoint with the following parameters:
+  <code>
+    /authorize?client_id=clientId&redirect_uri=mcp://callback&response_type=code&scope=scope&code_challenge=codeChallenge&code_challenge_method=S256
+  </code>
+
+  This endpoint will capture the parameters on `state` param and redirect to the upstream authorization server.
+*/
+authRouter.get(
+  '/authorize',
+  bodyParser.urlencoded({ extended: true }),
+  async (req: ExpressRequest, res: ExpressResponse) => {
+    const requestParams = parseAuthRequest(req);
+    const authUrl = await upstreamAuth(btoa(JSON.stringify(requestParams)));
+    res.redirect(authUrl.href);
+  },
+);
+
+/*
+  Handles the callback from the upstream authorization server and completes the authorization code grant flow with downstream MCP client.
+
+  Step 2:
+  Upstream authorization server will redirect to `/callback` with the authorization code.
+  <code>
+    /callback?code=authorizationCode&state=state
+  </code>
+
+  - Exchange the upstream authorization code for an access token.
+  - Generate new authorization code and grant id.
+  - Save the authorization code and access token in the database.
+  - Redirect to the MCP client with the new authorization code.
+*/
+authRouter.get(
+  '/callback',
+  bodyParser.urlencoded({ extended: true }),
+  async (req: ExpressRequest, res: ExpressResponse) => {
+    const tokens = await exchangeCode(req);
+    const state = req.query.state as string;
+    const requestParams = decodeAuthParams(state);
+
+    // Implicit grant or `response_type=token` is not supported
+    if (requestParams.responseType !== 'code') {
+      res
+        .status(400)
+        .json({ code: 'invalid_request', error: 'invalid response type' });
+      return;
+    }
+
+    const clientId = requestParams.clientId;
+    const client = await model.getClient(clientId, '');
+    if (!client) {
+      res
+        .status(400)
+        .json({ code: 'invalid_request', error: 'invalid client id' });
+      return;
+    }
+
+    // Standard authorization code grant
+    const grantId = generateRandomString(16);
+    const secret = generateRandomString(32);
```

Comment:
```suggestion
    const nonce = generateRandomString(32);
```

NIT: I'd call it nonce (number once) rather than a secret